components/apimgt/org.wso2.carbon.apimgt.rest.api.dcr/src/main/java/org/wso2/carbon/apimgt/rest/api/dcr/web/impl/RegistrationServiceImpl.java

```diff
@@ -167,4 +234,214 @@ private boolean isUserAccessAllowed(String username) {
+            ServiceProvider serviceprovider = new ServiceProvider();
+            serviceprovider.setApplicationName(applicationName);
+            serviceprovider.setDescription("Service Provider for application " + appName);
+            serviceprovider.setSaasApp(applicationInfo.getIsSaasApplication());
+
+            ApplicationManagementService appMgtService = ApplicationManagementService.getInstance();
+            appMgtService.createApplication(serviceprovider, tenantDomain, userName);
+
+            //retrieving the created service provider
+            ServiceProvider createdServiceProvider =
+                    appMgtService.getApplicationExcludingFileBasedSPs(applicationName, tenantDomain);
+            if (createdServiceProvider == null) {
+                throw new APIManagementException("Error occurred while creating Service Provider " +
+                        "Application" + appName);
+            }
+            //creating the OAuth app
+            OAuthConsumerAppDTO createdOauthApp =
+                    this.createOAuthApp(applicationName, callbackUrl, grantType, userName);
+
+            // Set the OAuthApp in InboundAuthenticationConfig
+            InboundAuthenticationConfig inboundAuthenticationConfig = new InboundAuthenticationConfig();
+            InboundAuthenticationRequestConfig[] inboundAuthenticationRequestConfigs =
+                    new InboundAuthenticationRequestConfig[1];
+            InboundAuthenticationRequestConfig inboundAuthenticationRequestConfig =
+                    new InboundAuthenticationRequestConfig();
+
+            String oAuthType = APIConstants.SWAGGER_12_OAUTH2;
+            inboundAuthenticationRequestConfig.setInboundAuthType(oAuthType);
+            inboundAuthenticationRequestConfig.setInboundAuthKey(createdOauthApp.getOauthConsumerKey());
+            String oauthConsumerSecret = createdOauthApp.getOauthConsumerSecret();
+
+            if (oauthConsumerSecret != null && !oauthConsumerSecret.isEmpty()) {
+                Property property = new Property();
+                property.setName(ApplicationConstants.INBOUNT_AUTH_CONSUMER_SECRET);
+                property.setValue(oauthConsumerSecret);
+                Property[] properties = {property};
+                inboundAuthenticationRequestConfig.setProperties(properties);
+            }
+
+            inboundAuthenticationRequestConfigs[0] = inboundAuthenticationRequestConfig;
+            inboundAuthenticationConfig.setInboundAuthenticationRequestConfigs(inboundAuthenticationRequestConfigs);
+            createdServiceProvider.setInboundAuthenticationConfig(inboundAuthenticationConfig);
+
+            //updating the service provider with Inbound Authentication Configs
+            appMgtService.updateApplication(createdServiceProvider, tenantDomain, userName);
+
+            Map<String, String> valueMap = new HashMap<String, String>();
+            valueMap.put(OAUTH_REDIRECT_URIS, createdOauthApp.getCallbackUrl());
```

Comment:
if there is no callback url, then should not allow to create app with authorization_code and implicit grant type.